File: src/schema/application/installation_codes.ts
import {
  AnyPgColumn,
  boolean,
  pgTable,
  primaryKey,
  text,
} from 'drizzle-orm/pg-core';
import {
  getConfigDefaults,
  system_fields,
} from '@dna-platform/crdt-lww-postgres/build/schema/system';
import { table as devices } from './devices';
import { sql } from 'drizzle-orm';

const table_name = 'installation_codes';

const fields = {
  device_id: text('device_id').references(
    () => devices.id as AnyPgColumn,
  ),
  device_code: text('device_code'),
  redeemed: boolean("redeemed").default(false),
  auto_authorization: boolean("auto_authorization").default(false)
}

const config = (table) => ({
  pk: primaryKey({ columns: [table.id] }),
  ...getConfigDefaults.defaultIndexes(table_name, table),
});

export const table = pgTable(
  table_name,
  {
    ...system_fields,
    ...fields,
    id: text('id'),
    token: text("token")
      .default(sql`substring(md5(random()::text) FROM 1 FOR 8)`)
      .unique(),
  },
  config,
);

File: bin/store/src/database/schema/tables/installation_codes.rs
use crate::define_table_schema;
use crate::generated::schema::generator::diesel_schema_definition::{
    types::*, DieselTableDefinition,
};
use crate::{system_fields, system_foreign_keys, system_indexes};

/**
 * Auto-converted from TypeScript schema source.
 */
pub struct InstallationCodesTable;

define_table_schema! {
    hypertable: false,
    fields: {
        // System fields - common across all tables ( REQUIRED )
        system_fields!(),

        device_id: nullable(text()),
        device_code: nullable(text()),
        redeemed: nullable(boolean()), default "false",
        auto_authorization: nullable(boolean()), default "false",
        token: nullable(text()), default "sql`substring(md5(random(",
    },
    indexes: {
        // System field indexes ( REQUIRED )
        system_indexes!("installation_codes"),
    },
    foreign_keys: {
        // System field foreign keys ( REQUIRED )
        system_foreign_keys!("installation_codes"),
        fk_installation_codes_device_id: { columns: ["device_id"], foreign_table: "devices", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
    }
}
